doc: add a proposal for foreign key (#35982)
pingcap/tidb: docs/design/2022-06-22-foreign-key.md

# Foreign Key Design Doc

- Author(s): [crazycs520](https://github.com/crazycs520)
- Tracking Issue: https://github.com/pingcap/tidb/issues/18209

## Abstract

This proposes an implementation of supporting foreign key constrain.

## DDL Technical Design

### Table Information Changes
Table's foreign key information will be stored in `model.TableInfo`:

```go
// TableInfo provides meta data describing a DB table.
type TableInfo struct {
	...
	ForeignKeys         []*FKInfo            `json:"fk_info"`
	ReferredForeignKeys []*ReferredFKInfo    `json:"cited_fk_info"`
	...
}

// FKInfo provides meta data describing a foreign key constraint.
type FKInfo struct {
    ID        int64       `json:"id"`
    Name      CIStr       `json:"fk_name"`

>     Enable    bool        `json:"enable"`

Reviewer: Change it to Version? There may be additional versions in the future.

- Does not currently support foreign keys for tables with user-defined partitioning. This includes both reference and child tables.
- A foreign key constraint cannot reference a virtual generated column, but stored generated column is ok.

#### Handle In DDL Owner

When DDL Owner handle create table job, DDL owner need to create a new table, and update the reference tables information.

At the same point in time, there may be two versions of the schema in the TiDB cluster, so we can't create new table and 
update all reference tables in one schema version, since this may break foreign key constrain, such as delete reference table
without foreign key constrain check in child table.

```sql
-- In TiDB-1 and Schema Version is 1
insert into t_has_foreign_key values (1, 1);

-- In TiDB-0 and  Schema Version is 0
delete from t_reference where id = 1; --Since doesn't know foreign key information in old version, so doesn't do foreign key constrain check.
```

So, when create a table with foreign key, we need multi-schema version change:

1. None -> Write Only: Create table with state is `write-only`, and update all reference tables info.

> 2. Write Only -> Done: Update the new created table state to `public`.

Author: 只有带有外键的 建表才是这个行为，否则和原来步骤一致，不影响建表速度。

When build `TableInfo`, we need to auto create an index for foreign key columns if there is no index cover foreign key columns.
And this is divides the problem into two cases:
- Case-1: No need to auto create index, and only add foreign key constrain.
- Case-2: Need auto create index for foreign key

#### Case-1: Only add foreign key constrain

The DDL owner handle add foreign key constrain step is:

1. None -> Write Only: add foreign key constrain which state is `write-only` into table and update the reference table info.
3. Write Only - Write Reorg: check all row in the table whether has related foreign key exists in reference table, we can use following SQL to check:
   ```sql
   select count(*) from t2 where t2.a not in (select id from t1);
   ```
   The expected result is `0`, otherwise, an error is returned and cancel the ddl job.
4. Write Reorg -> Public: update the foreign key constrain state to `public`.

DML should also check the foreign key constrain which state is `write-only`

#### Case-2: Auto create index for foreign key and add foreign key constrain

As TiDB support multi-schema change now, we can split this into 2 sub-ddl job.

> - Add Index DDL job

Reviewer 2: The index added here is public to use?

Author: Nice catch. Only when the added index is ready to use(finish add index reorg, but the state is not public), then we can start to add the foreign key constrain.

- Add Foreign Key Constrain DDL job

We should do add index DDL job first, after index ddl job finish `write-reorg` and ready for public, then start to do add foreign key constrain ddl job.

### Drop Table

If `foreign_key_checks` is `ON`, then drop the table which has foreign key references will be rejected.

```sql
> drop table t1;
(3730, "Cannot drop table 't1' referenced by a foreign key constraint 't2_ibfk_1' on table 't2'.")
```

### Drop Index

Drop index which used by foreign key will be rejected.

```sql
> set @@foreign_key_checks=0; -- Even disable foreign_key_checks, you still can't drop the index which used for foreign key constrain.
Query OK, 0 rows affected
> alter table t2 drop index fk;
(1553, "Cannot drop index 'fk': needed in a foreign key constraint")
```

### Rename Table/Column

Rename table which has foreign key references, should also need to update the related child table info.

```sql
create table t1 (id int key,a int, index(a));
create table t2 (id int key,a int, foreign key fk(a) references t1(id) ON DELETE CASCADE);

> rename table t1 to t11;

Reviewer 3: why not just let `FKInfo` only store the refered table id

Author: truncate table / modify column with reorg  -> change child table id.

Author: After trying, you are right, I think recorded table id and schema id are better here.

Author: em... We still need to record the table/schema name, not table/schema id. Here is an example:

```sql
mysql>set @@foreign_key_checks=0;
Query OK, 0 rows affected
Time: 0.000s
mysql>create table t2 (a int key, foreign key fk(a) references t1(id));
Query OK, 0 rows affected
Time: 0.011s
mysql>create table t1 (id int key);
Query OK, 0 rows affected
Time: 0.011s
mysql>set @@foreign_key_checks=1;
Query OK, 0 rows affected
Time: 0.000s
mysql>insert into t2 values (1);
(1452, 'Cannot add or update a child row: a foreign key constraint fails (`test`.`t2`, CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t1` (`id`))')
```

As you can see, the table `t2` is refer to table `t1`, and when create table `t2`, the table `t1` still not be created, so we can't know table `t1`.

    ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
```

### Modify Column

Modify column which used by foreign key will be rejected.

```sql
> alter table t1 change column id id1 bigint;
(3780, "Referencing column 'a' and referenced column 'id1' in foreign key constraint 't2_ibfk_1' are incompatible.")
```

MySQL modify column problem: https://www.percona.com/blog/2019/06/04/ddl-queries-foreign-key-columns-MySQL-pxc/

What if the user really need to modify column type, such as from `INT` to `BIGINT`. Maybe we can offer an variable such as `alter-foreign-keys-method=auto`,
then when user modify the column type, TiDB will auto modify the related foreign key column's type. For easy implementation and reduce risk, maybe only support modify column type which doesn't need to reorg table row data.

## DML Technical Design

### DML On Child Table

On Child Table Insert Or Update, need to Find FK column value whether exist in reference table:

1. Get reference table info by table name.
2. Get related fk index of reference table.
3. tiny optimize, check fk column value exist in reference table cache(map[string(index_key)]struct).
3. Get related row in reference.

>   - Construct index key and then use snapshot `Iter` and `Seek` API to scan. If the index is unique and only contain

Author: should add lock in parent table to avoid concurrent delete in parent table.

    foreign key columns, use snapshot `Get` API.
    - `Iter` default scan batch size is 256, need to set 2 to avoid read unnecessary data.
4. compact column value to make sure exist.
5. put column value into reference fk column value cache.

check order should check unique/primary key constrain first:

```sql
test> create table t1 (id int key,a int, index(a));
test> create table t2 (id int key,a int, foreign key fk(a) references t1(id) ON DELETE CASCADE);
test> insert into t1 values (1, 1);
test> insert into t2 values (1, 1);
test> insert into t2 values (1, 2);
(1062, "Duplicate entry '1' for key 't2.PRIMARY'")
test> insert ignore into t2 values (1, 2);
Query OK, 0 rows affected
```

Load data should also do foreign key check, but report warning instead error:

```sql
create table t1 (id int key,a int, index(a));

> create table t2 (id int key,a int, foreign key fk(a) references t1(id) ON DELETE CASCADE);

Reviewer 3: better add a placement rule here, let t1 & t2 which has the foreign-key bound ship table's leader scheduled in a single kv node.

Reviewer 3: I also think about the mechanism of a kind of inherited table means from postgre, letting table t2's encoded in the range of table t1's scope.

test> show warnings;
+---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Level   | Code | Message                                                                                                                                                           |
+---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Warning | 1452 | Cannot add or update a child row: a foreign key constraint fails (`test`.`t2`, CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t1` (`id`) ON DELETE CASCADE) |
| Warning | 1452 | Cannot add or update a child row: a foreign key constraint fails (`test`.`t2`, CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t1` (`id`) ON DELETE CASCADE) |
+---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

`data.csv` is:
```
1,1
2,2
```

### DML On reference Table

> On reference Table Insert Or Update:

Author: ```suggestion
On reference Table Delete Or Update:
```

- `CASCADE`: update/delete related child table row.
- `SET NULL`: set related child row's foreign key columns value to NULL.
- `RESTRICT`, `NO ACTION`: If related row exist in child table, reject update/delete reference table.
- `SET DEFAULT`: just like `RESTRICT`.

modify related child table row by following step:
1. get child table info by name(in reference table info).
2. get the child table fk index's column info.
3. build update executor to update child table rows.

### Issue need to be discussed

#### Affect Row

related article: https://www.percona.com/blog/hidden-cost-of-foreign-key-constraints-in-MySQL/

Here is a MySQL example:

prepare:
```sql
create table t1 (id int key,a int, index(a));
create table t2 (id int key,a int, foreign key fk(a) references t1(id) ON DELETE CASCADE);
insert into t1 values (1, 1);
insert into t2 values (1, 1);
```

Then delete on reference table:
```sql
> delete from t1 where id=1;
Query OK, 1 row affected
```

As you can see, in the query result, the affected row is 1, but actually should be 2, since the related row in t2 is also been deleted.

This is a MySQL issue, do we need to be compatible with it?

> 

Author: 需要 spec  写清楚这个行为，以及如何处理旧表中的 fk . @PM

  │                 KV rows read: 1
  │                 KV bytes read: 27 B
  │                 missing stats
  │                 table: customers_2@primary
  │                 spans: [/1 - /1]
  │                 locking strength: for update
  │
  └── • fk-cascade
        fk: fk_customer_id_ref_customers_2
        input: buffer 1
```

##### PostgreSQL DML Execution Plan

```sql
postgres=# explain analyze UPDATE customers_2 SET id = 20 WHERE id = 23;
                                                             QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------
 Update on customers_2  (cost=0.15..8.17 rows=1 width=10) (actual time=0.039..0.039 rows=0 loops=1)
   ->  Index Scan using customers_2_pkey on customers_2  (cost=0.15..8.17 rows=1 width=10) (actual time=0.016..0.016 rows=1 loops=1)
         Index Cond: (id = 23)
 Planning Time: 0.057 ms
 Trigger for constraint orders_2_customer_id_fkey on customers_2: time=0.045 calls=1
 Trigger for constraint orders_2_customer_id_fkey on orders_2: time=0.023 calls=2
 Execution Time: 0.129 ms
```

## Other Technical Design

> 

Author: 优化器利用 fk 做优化，例如 join

Author: plan 性能不能回退

### How to check foreign key integrity?

How MySQL to do this? Look like MySQL doesn't provide any method, but the user can use stored procedure to do this, see: https://stackoverflow.com/questions/2250775/force-innodb-to-recheck-foreign-keys-on-a-table-tables

Maybe We can use following syntax to check foreign key integrity:

```sql
ADMIN CHECK FOREIGN KEY [table_name] [foreign_key_name]
```

which implemention is use following SQL to check:

```sql
select count(*) from t where t.a not in (select id from t_refer);
```

## Impact

### Impact of data replication

> todo

Author: 导入数据需要关闭 fk check  @tool college.

Reviewer 4: @3pointer @overvenus PTAL

Query OK, 0 rows affected
mysql> show create table t\G
***************************[ 1. row ]***************************
Table        | t
Create Table | CREATE TABLE `t` (
  `id` int NOT NULL,
  `a` int DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `fk` (`a`),
  CONSTRAINT `t_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t` (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
1 row in set
```

As you can see, the index `fk` is created automatically for the foreign key.

#### Validate

Create a table with foreign key, check the following conditions when a DDL job is built and the DDL owner received a DDL job(aka Double-Check):

- whether the user have `REFERENCES` privilege to the foreign key references table.
- Corresponding columns in the foreign key and the referenced key must have similar data types. The size and sign of fixed precision types such as INTEGER and DECIMAL must be the same. The length of string types need not be the same. For nonbinary (character) string columns, the character set and collation must be the same.
- Supports foreign key references between one column and another within a table. (A column cannot have a foreign key reference to itself.)

> - Indexes on foreign keys and referenced keys are required, because the foreign key check can be fast and do not require a table scan. The size and sign of fixed precision types such as INTEGER and DECIMAL must be the same. The length of string types need not be the same. For nonbinary (character) string columns, the character set and collation must be the same.

Reviewer 5: ```suggestion
- Indexes on foreign keys and referenced keys are required, because the foreign key check can be fast and do not require a table scan.
```

Does it duplicate line 89?

- Prefixed indexes on foreign key columns are not supported. Consequently, BLOB and TEXT columns cannot be included in a foreign key because indexes on those columns must always include a prefix length.

> - Does not currently support foreign keys for tables with user-defined partitioning. This includes both reference and child tables.

Reviewer 5: What is "user-defined" partitioning?

Author: I change to the partition table.
`table with user-defined partitioning` is partition table.
<img width="537" alt="image" src="https://user-images.githubusercontent.com/26020263/207788983-968f1a7c-f9ba-4afa-b6e2-de404506b3a2.png">

- A foreign key constraint cannot reference any virtual generated columns, but the stored generated columns are fine.

#### Handle In DDL Owner

When DDL Owner handle create table job, DDL owner need to create a new table.

At the same point in time, there may be two versions of the schema in the TiDB cluster, so we can't create new table and 
update all reference tables in one schema version, since this may break foreign key constraint, such as delete reference table

> without foreign key constrain check in child table.

Reviewer 5: Please `s/constrain/constraint/g`

```sql
-- In TiDB-1 and Schema Version is 1
insert into t_has_foreign_key values (1, 1);

-- In TiDB-0 and  Schema Version is 0
delete from t_reference where id = 1; --Since doesn't know foreign key information in old version, so doesn't do foreign key constrain check.
```

So, when create a table with foreign key, we need multi-schema version change:

1. None -> Write Only: Create table with state is `write-only`.

> 2. Write Only -> Done: Update the created table state to `public`.

Reviewer 5: ```suggestion
2. Write Only -> Public: Update the created table state to `public`.
```

Then `applyTableUpdate` will also need to reload the old/new table's referred table information, also uses `applyDropTable` to drop the old reference table, use `applyCreateTable` to create new reference table.

That's all.

### Alter Table Add Foreign Key

Here is an example:

```sql
create table t1 (id int key,a int, index(a));
create table t2 (id int key,a int);
alter  table t2 add foreign key fk(a) references t1(id) ON DELETE CASCADE;
```

Just like create table, we should validate first, and return error if the conditions for creating foreign keys are not met, and also need to do double-check.

When build `TableInfo`, we need to auto create an index for foreign key columns if there is no index cover foreign key columns.
And this is divides the problem into two cases:
- Case-1: No need to create index automatically, and only add foreign key constraint.
- Case-2: Need auto create index for foreign key

#### Case-1: Only add foreign key constrain

The DDL owner handle add foreign key constrain step is:

1. None -> Write Only: add foreign key constrain which state is `write-only` into table.

> 3. Write Only - Write Reorg: check all row in the table whether has related foreign key exists in reference table, we can use following SQL to check:

Reviewer 5: ```suggestion
3. Write Only -> Write Reorg: check all row in the table whether has related foreign key exists in reference table, we can use following SQL to check:
```

   ```sql
   select 1 from t2 where t2.a is not null and t2.a not in (select id from t1) limit 1;
   ```
   The expected result is `empty`, otherwise, an error is returned and cancel the ddl job.
4. Write Reorg -> Public: update the foreign key constrain state to `public`.

A problem is, How the DML treat the foreign key with on delete/update cascade behaviour which state is non-public?
Here is an example:

```sql
create table t1 (id int key,a int, index(a));
create table t2 (id int key,a int, index(a));
insert into t1 values (1,1);
insert into t2 values (1,1);
alter  table t2 add constraint fk_1 foreign key (a) references t1(id) ON DELETE CASCADE;
```

The schema change of foreign key `fk_1` is from `None` -> `Write-Only` -> `Write-Reorg` -> `Public`。
When the foreign key `fk_1` in `Write-Only` state, a DML request has come to be processed:

```sql
delete from t1 where id = 1;
```

Then, TiDB shouldn't do cascade delete for foreign key `fk_1` in state `Write-Only`, since the `Add Foreign Key` DDL job maybe
failed in `Write-Reorg` state and rollback the DDL job. But it is hard to rollback the cascade deleted executed before.

> So, when execute DML with `non-publick` foreign key, TiDB will do foreign key constraint check instead of foreign key cascade behaviour.

Reviewer 5: ```suggestion
So, when execute DML with `non-public` foreign key, TiDB will do foreign key constraint check instead of foreign key cascade behaviour.
```

You can use `SHOW CREATE TABLE` to see whethere a foreign key is take effect, the old version foreign key will have a comment `/* FOREIGN KEY INVALID */` to indicate it is invalid, the new version foreign key doesn't have this comment.

```sql
> show create table t2\G
    ***************************[ 1. row ]***************************
    Table        | t2
Create Table | CREATE TABLE `t2` (
    `id` int(11) NOT NULL,
    PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,
    CONSTRAINT `fk` FOREIGN KEY (`id`) REFERENCES `test`.`t1` (`id`) ON DELETE CASCADE ON UPDATE CASCADE /* FOREIGN KEY INVALID */
    ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
```

## Impact

### Impact of data replication

#### TiCDC

When sync table data to downstream TiDB, TiCDC should `set @@foreign_key_checks=0` in downstream TiDB.

#### DM

When do full synchronization, DM should `set @@foreign_key_checks=0` in downstream TiDB.

When do incremental synchronization, DM should set `foreign_key_checks` session variable according to MySQL binlog.

#### BR

> When sync table data to downstream TiDB, TiCDC should `set @@foreign_key_checks=0` in downstream TiDB.

Reviewer 5: Title and content do not match.

  - `Version`: uses to distinguish between old and new versions.
Struct `ReferredFKInfo` uses for referenced table to record the child table which referenced me.

### Create Table with Foreign Key

#### Build TableInfo

When build `TableInfo`, auto create an index for foreign key columns if there is no index cover foreign key columns. Here is an example:

```sql
mysql>create table t (id int key,a int, foreign key fk(a) references t(id));
Query OK, 0 rows affected
mysql>show create table t\G
***************************[ 1. row ]***************************
Table        | t
Create Table | CREATE TABLE `t` (
  `id` int NOT NULL,
  `a` int DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `fk` (`a`),
  CONSTRAINT `t_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t` (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
1 row in set
```

As you can see, the index `fk` is auto create for foreign key usage.

#### Validate

Create a table with foreign key, check following condition when build DDL job and DDL owner received DDL job(aka Double-Check):

- whether the user have `REFERENCES` privilege to the foreign key references table.

> - Corresponding columns in the foreign key and the referenced key must have similar data types. The size and sign of fixed precision types such as INTEGER and DECIMAL must be the same. The length of string types need not be the same. For nonbinary (character) string columns, the character set and collation must be the same.

Author: MySQL use `similar` here.
![image](https://user-images.githubusercontent.com/26020263/183800711-07cf0c2b-a5c7-403a-aab9-dc1740d56c7d.png)

- A foreign key constraint cannot reference a virtual generated column, but stored generated column is ok.

#### Handle In DDL Owner

When DDL Owner handle create table job, DDL owner need to create a new table, and update the reference tables information.

At the same point in time, there may be two versions of the schema in the TiDB cluster, so we can't create new table and 
update all reference tables in one schema version, since this may break foreign key constrain, such as delete reference table
without foreign key constrain check in child table.

```sql
-- In TiDB-1 and Schema Version is 1
insert into t_has_foreign_key values (1, 1);

-- In TiDB-0 and  Schema Version is 0
delete from t_reference where id = 1; --Since doesn't know foreign key information in old version, so doesn't do foreign key constrain check.
```

So, when create a table with foreign key, we need multi-schema version change:

1. None -> Write Only: Create table with state is `write-only`, and update all reference tables info.
2. Write Only -> Done: Update the new created table state to `public`.

> In step-1, we need update some table info in one schema-version. Technically, we can implement is since we already support `ActionCreateTables` DDL job.

Reviewer 5: ```suggestion
In step-1, we need update some table info in one schema version. Technically, we can implement it since we already support `ActionCreateTables` DDL job.
```

### Alter Table Add Foreign Key

Here is an example:

```sql
create table t1 (id int key,a int, index(a));
create table t2 (id int key,a int);
alter  table t2 add foreign key fk(a) references t1(id) ON DELETE CASCADE;
```

Just like create table, we should validate first, and return error if the conditions for creating foreign keys are not met, and also need to do double-check.

When build `TableInfo`, we need to auto create an index for foreign key columns if there is no index cover foreign key columns.
And this is divides the problem into two cases:

> - Case-1: No need to auto create index, and only add foreign key constrain.

Reviewer 5: ```suggestion
- Case-1: No need to create index automatically, and only add foreign key constraint.
```

- Case-2: Need auto create index for foreign key

#### Case-1: Only add foreign key constrain

The DDL owner handle add foreign key constrain step is:

1. None -> Write Only: add foreign key constrain which state is `write-only` into table and update the reference table info.
3. Write Only - Write Reorg: check all row in the table whether has related foreign key exists in reference table, we can use following SQL to check:
   ```sql
   select count(*) from t2 where t2.a not in (select id from t1);
   ```
   The expected result is `0`, otherwise, an error is returned and cancel the ddl job.
4. Write Reorg -> Public: update the foreign key constrain state to `public`.

> 

Author: need concurrent test and rollback.

Author: need rollback design doc.

                                     `a` int DEFAULT NULL,
                                     PRIMARY KEY (`id`),
    KEY `fk` (`a`),
    CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t11` (`id1`) ON DELETE CASCADE
    ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
```

### Truncate Table

Truncate table which has foreign key or references, should also need to update the related child/parent table info.

### Modify Column

Modify column which used by foreign key will be rejected.

```sql
> alter table t1 change column id id1 bigint;
(3780, "Referencing column 'a' and referenced column 'id1' in foreign key constraint 't2_ibfk_1' are incompatible.")
```

MySQL modify column problem: https://www.percona.com/blog/2019/06/04/ddl-queries-foreign-key-columns-MySQL-pxc/

What if the user really need to modify column type, such as from `INT` to `BIGINT`. Maybe we can offer an variable such as `alter-foreign-keys-method=auto`,
then when user modify the column type, TiDB will auto modify the related foreign key column's type. For easy implementation and reduce risk, maybe only support modify column type which doesn't need to reorg table row data.

> 

Author: Alter index invisible need check too.

-------------------------------------------------------------------------------------------------------------------------------------
 Update on customers_2  (cost=0.15..8.17 rows=1 width=10) (actual time=0.039..0.039 rows=0 loops=1)
   ->  Index Scan using customers_2_pkey on customers_2  (cost=0.15..8.17 rows=1 width=10) (actual time=0.016..0.016 rows=1 loops=1)
         Index Cond: (id = 23)
 Planning Time: 0.057 ms
 Trigger for constraint orders_2_customer_id_fkey on customers_2: time=0.045 calls=1
 Trigger for constraint orders_2_customer_id_fkey on orders_2: time=0.023 calls=2
 Execution Time: 0.129 ms
```

## Other Technical Design

### How to check foreign key integrity?

How MySQL to do this? Look like MySQL doesn't provide any method, but the user can use stored procedure to do this, see: https://stackoverflow.com/questions/2250775/force-innodb-to-recheck-foreign-keys-on-a-table-tables

Maybe We can use following syntax to check foreign key integrity:

```sql
ADMIN CHECK FOREIGN KEY [table_name] [foreign_key_name]
```

which implemention is use following SQL to check:

```sql
select count(*) from t where t.a not in (select id from t_refer);

> ```

Reviewer 5: I don't think a new syntax is necessary if there is no optimization.. `select count(*)` is fine.

    if err != nil || hasNullValue(vals) {
        return err
    }
    fkt.fkValues = append(fkt.fkValues, vals)
    return nil
}
```

4. Pseudocode after `Delete` execution.

```go
func (a *ExecStmt) Exec(){
    e, err := a.buildExecutor()
    e.Open()
	f handled, result, err := a.handleNoDelay(ctx, e, isPessimistic); handled {
        err = a.handleForeignKeyTrigger(ctx, e, isPessimistic)
        return result, err
    }
}

func (a *ExecStmt) handleForeignKeyTrigger(ctx context.Context, e Executor, isPessimistic bool) error {
    fkTriggerExecs := e.GetForeignKeyTriggerExecs()
    for i := 0; i < len(fkTriggerExecs); i++{
        fkt := fkTriggerExecs[i]
        e, err := fkt.buildExecutor(ctx)
        if err != nil {
            return err
        }
        if e == nil {
            continue
        }
        if err := e.Open(ctx); err != nil {
            terror.Call(e.Close)
            return err
        }
        _, _, err = a.handleNoDelay(ctx, e, isPessimistic)
        if err != nil {
            return err
        }
        err = a.handleForeignKeyTrigger(ctx, e, isPessimistic)
        if err != nil {
            return err
        }
    }
    return nil
}
```

> ### Issue need to be discussed

Reviewer: Do we need to add https://github.com/pingcap/tidb/pull/37148#discussion_r957951003 to here?

Author: done.

| Level   | Code | Message                                                                                                                                                           |
+---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Warning | 1452 | Cannot add or update a child row: a foreign key constraint fails (`test`.`t2`, CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t1` (`id`) ON DELETE CASCADE) |
| Warning | 1452 | Cannot add or update a child row: a foreign key constraint fails (`test`.`t2`, CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t1` (`id`) ON DELETE CASCADE) |
+---------+------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

`data.csv` is:
```
1,1
2,2
```

### DML On reference Table

On reference Table Delete Or Update:

> 1. modify related child table row by referential action:

Reviewer 4: Can these actions also be turned off by the session variable?

 Update on customers_2  (cost=0.15..8.17 rows=1 width=10) (actual time=0.039..0.039 rows=0 loops=1)
   ->  Index Scan using customers_2_pkey on customers_2  (cost=0.15..8.17 rows=1 width=10) (actual time=0.016..0.016 rows=1 loops=1)
         Index Cond: (id = 23)
 Planning Time: 0.057 ms
 Trigger for constraint orders_2_customer_id_fkey on customers_2: time=0.045 calls=1
 Trigger for constraint orders_2_customer_id_fkey on orders_2: time=0.023 calls=2
 Execution Time: 0.129 ms
```

## Other Technical Design

### How to check foreign key integrity?

How MySQL to do this? Look like MySQL doesn't provide any method, but the user can use stored procedure to do this, see: https://stackoverflow.com/questions/2250775/force-innodb-to-recheck-foreign-keys-on-a-table-tables

Maybe We can use following syntax to check foreign key integrity:

```sql
ADMIN CHECK FOREIGN KEY [table_name] [foreign_key_name]
```

which implemention is use following SQL to check:

```sql
select count(*) from t where t.a not in (select id from t_refer);
```

## Impact

### Impact of data replication

todo

> 

Reviewer 4: For one-time data load/import/restore, if the implementation is based on SQL layer (for example, lightning tidb backend), the behaviour should be the same as the behavour in "DML Load data" section. If it's based on storage later (for example, lightning local backend), we do not guarantee any foreign key properties, including:
- inserted rows has no respect for foreign key, as `set @@foreign_key_checks=0`
- overwritten rows will also not trigger referenced table actions as in section "DML On reference Table"

Reviewer 4: For long-time running data replication like CDC/DM, I can list some replication standards in the order of enabling more functionality:

1. mask out FK features when replicating
2. replicate FK related statements as-is, but in downstream runtime disable FK features
3. the FK features is correctly enabled during replicating

Seems currently standard level 2 is easy to accomplish, and the shortcoming is DML on referenced table will not trigger ["Referential Actions"](https://dev.mysql.com/doc/refman/8.0/en/create-table-foreign-keys.html#foreign-key-referential-actions) like DELETE CASCADE or SET NULL. Is it acceptable? @frank945946

And please help us check if these criterias can meet @crazycs520 
- There's a way to execute DDL with FK just like it does not contain FK. Specifically, can drop the index required by a foreign key
- There's a way to do DML on child tables like it does not contain FK
- There's a way to do DML on referenced table like it does not contain FK

Author: >Seems currently standard level 2 is easy to accomplish, and the shortcoming is DML on referenced table will not trigger ["Referential Actions"](https://dev.mysql.com/doc/refman/8.0/en/create-table-foreign-keys.html#foreign-key-referential-actions) like DELETE CASCADE or SET NULL. Is it acceptable?

For data replication like CDC, since CDC watch the kv change then generage SQL to to data replication in downstream databases, so disable FK feature in downstream is ok, since if upstream enable FK feature and does Referential Actions, CDC will generate addiational SQL about it.

For data replication like DM, I think we should confirm the binlog of DML with FK, 
* if the binlog contains the FK Referential Actions, then disable FK feature in downstream is ok and generate additional SQL executed in downstream.
*  If the binlog doesn't contain the FK Referential Actions, then downstream maybe should enable FK feature.

Author: >And please help us check if these criterias can meet @crazycs520
>- There's a way to execute DDL with FK just like it does not contain FK. Specifically, can drop the index required by a foreign key
>- There's a way to do DML on child tables like it does not contain FK
>- There's a way to do DML on referenced table like it does not contain FK

Mostly, after setting `set @@foreign_key_checks=0`, mostly DDL and all DML will action like it doesn't contain FK.

A special case is `DROP INDEX` DDL, if the index is used by FK, then you can't drop the index no matter how.
